packages/workflow/src/internals.ts

```diff
@@ -157,38 +160,76 @@ export class Activator implements ActivationHandler {
   /**
    * Mapping of query name to handler
    */
-  public readonly queryHandlers = new Map<string, WorkflowQueryType>([
+  public readonly queryHandlers = new Map<string, { handler: WorkflowQueryType; description?: string }>([
     [
       '__stack_trace',
-      () => {
-        return this.getStackTraces()
-          .map((s) => s.formatted)
-          .join('\n\n');
+      {
+        handler: () => {
+          return this.getStackTraces()
+            .map((s) => s.formatted)
+            .join('\n\n');
+        },
+        description: 'Returns a sensible stack trace.',
       },
     ],
     [
       '__enhanced_stack_trace',
-      (): EnhancedStackTrace => {
-        const { sourceMap } = this;
-        const sdk: SDKInfo = { name: 'typescript', version: pkg.version };
-        const stacks = this.getStackTraces().map(({ structured: locations }) => ({ locations }));
-        const sources: Record<string, FileSlice[]> = {};
-        if (this.showStackTraceSources) {
-          for (const { locations } of stacks) {
-            for (const { filePath } of locations) {
-              if (!filePath) continue;
-              const content = sourceMap?.sourcesContent?.[sourceMap?.sources.indexOf(filePath)];
-              if (!content) continue;
-              sources[filePath] = [
-                {
-                  content,
-                  lineOffset: 0,
-                },
-              ];
+      {
+        handler: (): EnhancedStackTrace => {
+          const { sourceMap } = this;
+          const sdk: SDKInfo = { name: 'typescript', version: pkg.version };
+          const stacks = this.getStackTraces().map(({ structured: locations }) => ({ locations }));
+          const sources: Record<string, FileSlice[]> = {};
+          if (this.showStackTraceSources) {
+            for (const { locations } of stacks) {
+              for (const { filePath } of locations) {
+                if (!filePath) continue;
+                const content = sourceMap?.sourcesContent?.[sourceMap?.sources.indexOf(filePath)];
+                if (!content) continue;
+                sources[filePath] = [
+                  {
+                    content,
+                    lineOffset: 0,
+                  },
+                ];
+              }
             }
           }
-        }
-        return { sdk, stacks, sources };
+          return { sdk, stacks, sources };
+        },
+        description: 'Returns a stack trace annotated with source information.',
+      },
+    ],
+    [
+      '__temporal_workflow_metadata',
+      {
+        handler: (): temporal.api.sdk.v1.IWorkflowMetadata => {
+          const workflowType = this.info.workflowType;
+          const description = this.info?.memo?.__temporal_workflow_description;
```

Comment:
Hmm... I wasn't part of the design but why put this in the memo? And if we're putting this in the memo, why also return it here?

Reply:
I think we should question if we want to put this in memo since we now have the standard query (although I see some advantages to it such as not needing a worker up to get the description).

If we want to use both memo and query, it might be better to have a dedicated method to set the description for a workflow.

Reply:
I'm a bit reluctant with configuring this through a memo. At present, given that we don't support `upsertMemo`, that means the description would have to be provided by the code scheduling the workflow (i.e. client or parent workflow), which isn't desirable. Or, if we have `upsertMemo`, description would have to be set after starting the workflow, through a command, making this a billable feature.

Reply:
I don't really want to change WorkflowExecutionInfo protos for something that is going to be rarely used, and its functionality is redundant to memo. And it has to work across continueAsNew, or multiple workers, so I cannot see how to do that without a proto change.

I'm inclined to just not set it at all, and if someone needs extra workflow info they can use memo + lookup with the  workflow id. Alternatively, we could do an explicit setter as @bergundy mentioned, after upsertMemo is implemented in TS. 

About the cost of upsertMemo, I don't think the WF description  is needed to use the query method, today the UI just uses the handlers names, and, in any case, the query costs are going to dominate anyway...

Any thoughts @bergundy @mjameswh ?

Reply:
Memo is not a good place to put this IMO. Memo takes up visibility store room and is only for information that needs to be returned on list/describe.

> I don't really want to change WorkflowExecutionInfo protos for something that is going to be rarely used, and its functionality is redundant to memo.

I don't think this information is related to existing protos and such. This is just a query return value. If there is a TS-SDK-specific problem of being able to set metadata for a workflow, ok, but this issue should be _just_ about returning this information from query and nothing more IMO.

Reply:
@cretz I was following your proto schema that included a workflow description. What kind of workflow description you don't want to appear in list/describe? Do we need this at all?
I though that in all SDKs you need to modify protos if you want some workflow info to propagate across continueAsNew and worker's multiple execution resume. Is there a standard way across SDKs to do that without modifying protos?

Reply:
> I was following your proto schema that included a workflow description
Which proto schema? If you mean https://github.com/temporalio/api/pull/336, that was meant purely as a query result and those messages aren't used otherwise (yet).

> What kind of workflow description you don't want to appear in list/describe? Do we need this at all?

I wasn't expecting list/describe to be affected by this at all. Those are server-side calls. Maybe one day, but for now this is just querying a running workflow for information about itself at this time (sorry if there was confusion on the feature repo issue, it got a bit scatted).

>  I though that in all SDKs you need to modify protos if you want some workflow info to propagate across continueAsNew and worker's multiple execution resume. Is there a standard way across SDKs to do that without modifying protos?

This is fixed code-based metadata, it is not runtime data necessarily. You should just answer the query with the information the workflow code knows about itself and nothing more.

Reply:
>This is fixed code-based metadata, it is not runtime data necessarily. You should just answer the query with the >information the workflow code knows about itself and nothing more.

OK, that changes things, I thought it could change per workflow instance and not only per workflow type.

I don't know a clean way to decorate a function in a way that works both for TS and JS. We could patch the function object to add a static method/getter, or define a static map mapping function names to descriptions, but they both look like a hack... 

We could also just return the function name in the description for now...

Reply:
> I thought it could change per workflow instance and not only per workflow type.

It can change per workflow instance. People can manually register signal handlers for instance. But for the vast majority it won't change per instance. As for workflow description, I don't know of a good way for TS SDK to provide this kind of metadata and I'd be ok leaving it unset until decided.

> I don't know a clean way to decorate a function in a way that works both for TS and JS. We could patch the function object to add a static method/getter, or define a static map mapping function names to descriptions, but they both look like a hack...

Same. Users could call `setDescription` inside the workflow at runtime or somehow at registration time, but in the meantime I'm ok leaving this unset for TS SDK while we think about it.

> We could also just return the function name in the description for now...

I vote leave it unset